Fixes #373 allowing support for parents to use <content> or <slot>

admwx7 opened the pull request:
Updated the query to get the <template> element to use the effective children API which supports passing along the <template> from outside a wrapping element via <slot> or <template>

@blasten Fixes #373

blasten commented:
Thanks for the PR @admwx7! There have been attempts to solve this issue in the past https://github.com/PolymerElements/iron-list/pull/314.  That said, the change looks good. There are a few limitations that we will need to document. I'll take a look at them tomorrow.

admwx7 commented:
What limitations did you notice @blasten?  I haven't found any yet and I've been running this fix for around a day now.

My current demo structure uses a `<hc-nested-list-demo>` element which contains the following:
style
```
.item {
  cursor: pointer;
}
.item:hover, .item[open] {
  font-weight: bold;
}
```
template
```
<hc-nested-list items="{{items}}">
  <template>
    <div class="item" open$="[[item.open]]">
      <hc-nested-list-item item="{{item}}">
        <span>[[item.label]]</span>
      </hc-nested-list-item>
    </div>
  </template>
</hc-nested-list>
```
`<hc-nested-list>` uses the following template
```
<hc-nested-list items="{{items}}">
  <template>
    <div class="item" open$="[[item.open]]">
      <hc-nested-list-item item="{{item}}">
        <span>[[item.label]]</span>
      </hc-nested-list-item>
    </div>
  </template>
</hc-nested-list>
```
I have a click handler in `<hc-nested-list-item>` that fires an event for `<hc-nested-list>` to handle for grabbing the item's children then injecting them into the list to emulate nesting, it then reflects the "open" state as an attribute onto the item and the bold styling gets applied as expected.

The only limitation I've had to work around so far was in place before my fix was, which is that if I update an items state outside of it's template (updating the items array) then already rendered items state doesn't update, I handled this with a callback function inside of `<hc-nested-list-item>` that triggers an update to it's version of the item when the outside version is updated.

blasten commented:
@admwx7 Thanks! The limitation is that two-way binding to a parent property won't work, but now I just realized that we never handled that case.

admwx7 commented:
If that's the case @blasten can we get this merged in, I might take a stab at 2-way bindings in another PR at some point, but that would be a new feature, this is more of a bug fix.

blasten commented on iron-list.html:
```diff
@@ -1028,7 +1028,7 @@
         props.tabIndex = true;
         this._instanceProps = props;
-        this._userTemplate = Polymer.dom(this).querySelector('template');
+        this._userTemplate = this.queryEffectiveChildren('template');
```
One last thing, in L922, the physical items are appended as direct children:
```js
Polymer.dom(this).appendChild(inst.root);
```
Let's change that to:
```js
Polymer.dom(this._userTemplate.parentNode).appendChild(inst.root);
```
That way, users can style the template from where the template is defined.

admwx7 commented:
@blasten that should do the trick, added your change and it broke a few tests, had to go through and make a few other tweaks, they are pretty straight forward, except for the changes to focusing an item, the tests are passing and it works as expected in my usages (tested via `bower link`) but it's not exactly a straight forward fix.

blasten commented on iron-list.html:
```diff
@@ -1619,7 +1619,6 @@
       }
       var modelTabIndex, activeElTabIndex;
       var target = Polymer.dom(e).path[0];
-      var activeEl = Polymer.dom(this.domHost ? this.domHost.root : document).activeElement;
```
this shouldn't be removed. It could be:  
```js
var itemsHost = this._userTemplate.parentNode.domHost;
var activeEl = Polymer.dom(itemsHost || document).activeElement;
```

blasten commented on test/template-overload.html:
```diff
@@ -0,0 +1,103 @@
+</head>
+<body>
+
+  <test-fixture id="defaultTemplateList">
```
we can remove this test-fixture.

blasten commented on test/template-overload.html:
```diff
@@ -0,0 +1,103 @@
+      });
+    });
+
+    suite('default template', function() {
```
Replace to:

```js
suite('default template', function() {
  var list;

  setup(function() {
    list = fixture('templateOverloadList');
  });

  test('check physical item size', function(done) {
    var setSize = 10;
    list.items = buildDataSet(setSize);

    flush(function() {
      assert.equal(list.items.length, setSize);
      done();
    });
  });

  test('check count of physical items', function(done) {
    var setSize = 1;
    list.items = buildDataSet(setSize);

    flush(function() {
      assert.equal(Polymer.dom(list).querySelectorAll('*').length - 1, setSize);
      done();
    });
  });
});
```

admwx7 replied:
This will make both test suites test against the <slot> being overridden, should I just add that count test to the other suite and remove this one?

blasten commented on iron-list.html:
```diff
@@ -1732,7 +1726,7 @@
     _removeFocusedItem: function() {
       if (this._offscreenFocusedItem) {
-        Polymer.dom(this).removeChild(this._offscreenFocusedItem);
+        Polymer.dom(this._userTemplate.parentNode).removeChild(this._offscreenFocusedItem);
```
`this._itemsParent.removeChild(node)`

admwx7 replied:
I assume by `node` you actually mean `this._offscreenFocusedItem` here since `node` isn't defined.

blasten commented on iron-list.html:
```diff
@@ -919,7 +919,7 @@
         // on a doc fragment.
         physicalItems[i] = inst.root.querySelector('*');
-        Polymer.dom(this).appendChild(inst.root);
+        Polymer.dom(this._userTemplate.parentNode).appendChild(inst.root);
```
`this._itemsParent.appendChild`. Also in L1759: `this._itemsParent.appendChild(stampedTemplate.root);`

admwx7 commented:
Made the requested updates @blasten went ahead and squashed them together into one commit message to clean it up a bit while I was addressing this last round of changes.  The only thing that I wasn't certain of was the changes you wanted me to make to the test file, so I went ahead and just reduced it down to one test suite and added in the new test statement you had.

blasten commented:
@admwx7 Thanks for the changes, this is _really_ awesome!
